File: tests/downmix_test.py
#!/usr/bin/env python3
"""
Spatial Downmix Engine - Standalone Test Harness
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src', 'downmix'))

from spatial_downmix import (
    SpatialDownmix,
    SpatialDownmixConfig,
    SpatialLayout,
    SpatialMatrixMode,
    SpatialContentType,
    get_default_config_5_1,
    get_default_config_7_1,
)

import math


SAMPLE_RATE = 48000
DURATION_SEC = 1.0
NUM_FRAMES = int(SAMPLE_RATE * DURATION_SEC)


def analyze_stereo(data):
    sum_l = 0.0
    sum_r = 0.0
    peak_l = 0.0
    peak_r = 0.0
    clip_l = 0
    clip_r = 0

    for i in range(0, len(data), 2):
        l = data[i]
        r = data[i + 1]
        sum_l += l * l
        sum_r += r * r
        if abs(l) > peak_l:
            peak_l = abs(l)
        if abs(r) > peak_r:
            peak_r = abs(r)
        if abs(l) >= 1.0:
            clip_l += 1
        if abs(r) >= 1.0:
            clip_r += 1

    frames = len(data) // 2
    return {
        'rms_left': math.sqrt(sum_l / frames) if frames > 0 else 0.0,
        'rms_right': math.sqrt(sum_r / frames) if frames > 0 else 0.0,
        'peak_left': peak_l,
        'peak_right': peak_r,
        'clipping_left': clip_l,
        'clipping_right': clip_r,
    }


def print_analysis(label, a):
    print(f"{label}:")
    print(f"  RMS:   L={a['rms_left']:.6f}  R={a['rms_right']:.6f}")
    print(f"  Peak:  L={a['peak_left']:.6f}  R={a['peak_right']:.6f}")
    print(f"  Clip:  L={a['clipping_left']}  R={a['clipping_right']}")


def generate_tone(frames, channel, num_channels, freq=440.0, amplitude=1.0):
    data = [0.0] * (frames * num_channels)
    for i in range(frames):
        t = i / SAMPLE_RATE
        sample = amplitude * math.sin(2.0 * math.pi * freq * t)
        data[i * num_channels + channel] = sample
    return data


def test_isolated_channel_5_1(channel, name, gain):
    print(f"\n=== Test: Isolated {name} (gain={gain:.2f}) ===")

    input_data = generate_tone(NUM_FRAMES, channel, 6, 440.0, 1.0)

    config = get_default_config_5_1()
    config.debug_enabled = True

    if channel == 0:
        config.gains_5_1.left = gain
    elif channel == 1:
        config.gains_5_1.right = gain
    elif channel == 2:
        config.gains_5_1.center = gain
    elif channel == 3:
        config.gains_5_1.lfe = gain
    elif channel == 4 or channel == 5:
        config.gains_5_1.surround = gain

    downmix = SpatialDownmix(config)
    output = downmix.process_interleaved(input_data, NUM_FRAMES)

    a = analyze_stereo(output)
    print_analysis("Output", a)

    return a


def test_5_1_defaults():
    print("\n=== Test: 5.1 Default Configuration ===")

    input_data = [0.0] * (NUM_FRAMES * 6)

    config = get_default_config_5_1()
    config.debug_enabled = True

    downmix = SpatialDownmix(config)
    output = downmix.process_interleaved(input_data, NUM_FRAMES)

    a = analyze_stereo(output)
    print_analysis("Silence Output", a)


def test_stereo_passthrough():
    print("\n=== Test: Stereo Passthrough ===")

    input_data = [0.0] * (NUM_FRAMES * 2)
    for i in range(NUM_FRAMES):
        t = i / SAMPLE_RATE
        input_data[i * 2 + 0] = 0.5 * math.sin(2.0 * math.pi * 440.0 * t)
        input_data[i * 2 + 1] = 0.5 * math.sin(2.0 * math.pi * 550.0 * t)

    config = get_default_config_5_1()
    config.layout = SpatialLayout.STEREO

    downmix = SpatialDownmix(config)
    output = downmix.process_interleaved(input_data, NUM_FRAMES)

    a = analyze_stereo(output)
    print_analysis("Stereo Output", a)


def test_live_config_update():
    print("\n=== Test: Live Config Update ===")

    input_data = generate_tone(NUM_FRAMES, 2, 6, 440.0, 1.0)

    config = get_default_config_5_1()
    downmix = SpatialDownmix(config)

    config1 = get_default_config_5_1()
    config1.gains_5_1.center = 0.7
    downmix.update_config(config1)
    output1 = downmix.process_interleaved(input_data, NUM_FRAMES)
    a1 = analyze_stereo(output1)
    print_analysis("center=0.7", a1)

    config2 = get_default_config_5_1()
    config2.gains_5_1.center = 1.0
    downmix.update_config(config2)
    output2 = downmix.process_interleaved(input_data, NUM_FRAMES)
    a2 = analyze_stereo(output2)
    print_analysis("center=1.0", a2)

    config3 = get_default_config_5_1()
    config3.gains_5_1.center = 1.3
    downmix.update_config(config3)
    output3 = downmix.process_interleaved(input_data, NUM_FRAMES)
    a3 = analyze_stereo(output3)
    print_analysis("center=1.3", a3)

    ratio_1_0 = a2['rms_left'] / a1['rms_left'] if a1['rms_left'] > 0 else 0
    expected = 1.0 / 0.7
    print(f"Gain ratio (1.0/0.7): measured={ratio_1_0:.4f} expected={expected:.4f} diff={abs(ratio_1_0 - expected):.4f}")


def test_7_1_defaults():
    print("\n=== Test: 7.1 Default Configuration ===")

    input_data = [0.0] * (NUM_FRAMES * 8)

    config = get_default_config_7_1()
    config.debug_enabled = True

    downmix = SpatialDownmix(config)
    output = downmix.process_interleaved(input_data, NUM_FRAMES)

    a = analyze_stereo(output)
    print_analysis("Silence Output", a)


def test_7_1_isolated():
    print("\n=== Test: 7.1 Isolated Channels ===")

    channels = [
        (0, "FL", "left"),
        (1, "FR", "right"),
        (2, "C", "center"),
        (3, "LFE", "lfe"),
        (4, "SL", "side"),
        (5, "SR", "side"),
        (6, "BL", "rear"),
        (7, "BR", "rear"),
    ]

    for ch, name, attr in channels:
        config = get_default_config_7_1()
        config.debug_enabled = True
        setattr(config.gains_7_1, attr, 1.0)

        input_data = generate_tone(NUM_FRAMES, ch, 8, 440.0, 1.0)
        downmix = SpatialDownmix(config)
        output = downmix.process_interleaved(input_data, NUM_FRAMES)

        a = analyze_stereo(output)
        print_analysis(f"{name} (gain=1.0)", a)


def test_matrix_mode_none():
    print("\n=== Test: Matrix Mode NONE (vanilla) ===")

    config = get_default_config_5_1()
    config.matrix_oba = SpatialMatrixMode.NONE
    config.matrix_cba = SpatialMatrixMode.NONE
    config.debug_enabled = True

    downmix = SpatialDownmix(config)
    print(f"Matrix OBA: {config.matrix_oba.name}")
    print(f"Matrix CBA: {config.matrix_cba.name}")


def main():
    print("=" * 60)
    print("Spatial Downmix Engine - Standalone Test Harness")
    print("=" * 60)

    test_5_1_defaults()
    test_stereo_passthrough()

    test_isolated_channel_5_1(0, "FL (Front Left)", 1.0)
    test_isolated_channel_5_1(1, "FR (Front Right)", 1.0)
    test_isolated_channel_5_1(2, "C  (Center)", 1.0)
    test_isolated_channel_5_1(3, "LFE", 1.0)
    test_isolated_channel_5_1(4, "SL (Side Left)", 1.0)
    test_isolated_channel_5_1(5, "SR (Side Right)", 1.0)

    test_live_config_update()

    test_7_1_defaults()
    test_7_1_isolated()

    test_matrix_mode_none()

    print("\n" + "=" * 60)
    print("All Tests Complete")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: tests/downmix_test.c
#include "spatial_downmix.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define SAMPLE_RATE 48000
#define DURATION_SEC 1.0
#define NUM_FRAMES (SAMPLE_RATE * DURATION_SEC)

typedef struct {
    float rms_left;
    float rms_right;
    float peak_left;
    float peak_right;
    int clipping_left;
    int clipping_right;
} analysis_t;

static void analyze_stereo(const float* data, size_t frames, analysis_t* out) {
    float sum_l = 0, sum_r = 0;
    float peak_l = 0, peak_r = 0;
    int clip_l = 0, clip_r = 0;

    for (size_t i = 0; i < frames; i++) {
        float l = data[i * 2];
        float r = data[i * 2 + 1];
        sum_l += l * l;
        sum_r += r * r;
        if (fabsf(l) > peak_l) peak_l = fabsf(l);
        if (fabsf(r) > peak_r) peak_r = fabsf(r);
        if (fabsf(l) >= 1.0f) clip_l++;
        if (fabsf(r) >= 1.0f) clip_r++;
    }

    out->rms_left = sqrtf(sum_l / frames);
    out->rms_right = sqrtf(sum_r / frames);
    out->peak_left = peak_l;
    out->peak_right = peak_r;
    out->clipping_left = clip_l;
    out->clipping_right = clip_r;
}

static void print_analysis(const char* label, const analysis_t* a) {
    printf("%s:\n", label);
    printf("  RMS:   L=%.6f  R=%.6f\n", a->rms_left, a->rms_right);
    printf("  Peak:  L=%.6f  R=%.6f\n", a->peak_left, a->peak_right);
    printf("  Clip:  L=%d  R=%d\n", a->clipping_left, a->clipping_right);
}

static void generate_tone(float* buf, size_t frames, int channel, float freq, float amplitude) {
    for (size_t i = 0; i < frames; i++) {
        float t = (float)i / SAMPLE_RATE;
        float sample = amplitude * sinf(2.0f * M_PI * freq * t);
        for (int c = 0; c < 6; c++) {
            buf[i * 6 + c] = 0.0f;
        }
        buf[i * 6 + channel] = sample;
    }
}

static void test_isolated_channel_5_1(int channel, const char* name, float gain) {
    printf("\n=== Test: Isolated %s (gain=%.2f) ===\n", name, gain);

    float* input = malloc(NUM_FRAMES * 6 * sizeof(float));
    float* output = malloc(NUM_FRAMES * 2 * sizeof(float));

    generate_tone(input, NUM_FRAMES, channel, 440.0f, 1.0f);

    spatial_downmix_config_t config;
    spatial_downmix_get_default_config_5_1(&config);
    config.debug_enabled = 1;

    switch (channel) {
        case 0: config.gains_5_1.left = gain; break;
        case 1: config.gains_5_1.right = gain; break;
        case 2: config.gains_5_1.center = gain; break;
        case 3: config.gains_5_1.lfe = gain; break;
        case 4: config.gains_5_1.surround = gain; break;
        case 5: config.gains_5_1.surround = gain; break;
    }

    spatial_downmix_ctx_t* ctx = spatial_downmix_create(&config);
    spatial_downmix_process(ctx, input, output, NUM_FRAMES);
    spatial_downmix_destroy(ctx);

    analysis_t a;
    analyze_stereo(output, NUM_FRAMES, &a);
    print_analysis("Output", &a);

    free(input);
    free(output);
}

static void test_5_1_defaults() {
    printf("\n=== Test: 5.1 Default Configuration ===\n");

    float* input = malloc(NUM_FRAMES * 6 * sizeof(float));
    float* output = malloc(NUM_FRAMES * 2 * sizeof(float));

    for (size_t i = 0; i < NUM_FRAMES; i++) {
        for (int c = 0; c < 6; c++) {
            input[i * 6 + c] = 0.0f;
        }
    }

    spatial_downmix_config_t config;
    spatial_downmix_get_default_config_5_1(&config);
    config.debug_enabled = 1;

    spatial_downmix_ctx_t* ctx = spatial_downmix_create(&config);
    spatial_downmix_process(ctx, input, output, NUM_FRAMES);
    spatial_downmix_destroy(ctx);

    analysis_t a;
    analyze_stereo(output, NUM_FRAMES, &a);
    print_analysis("Silence Output", &a);

    free(input);
    free(output);
}

static void test_stereo_passthrough() {
    printf("\n=== Test: Stereo Passthrough ===\n");

    float* input = malloc(NUM_FRAMES * 2 * sizeof(float));
    float* output = malloc(NUM_FRAMES * 2 * sizeof(float));

    for (size_t i = 0; i < NUM_FRAMES; i++) {
        float t = (float)i / SAMPLE_RATE;
        input[i * 2 + 0] = 0.5f * sinf(2.0f * M_PI * 440.0f * t);
        input[i * 2 + 1] = 0.5f * sinf(2.0f * M_PI * 550.0f * t);
    }

    spatial_downmix_config_t config;
    spatial_downmix_get_default_config_5_1(&config);
    config.layout = SPATIAL_LAYOUT_STEREO;

    spatial_downmix_ctx_t* ctx = spatial_downmix_create(&config);
    spatial_downmix_process(ctx, input, output, NUM_FRAMES);
    spatial_downmix_destroy(ctx);

    analysis_t a;
    analyze_stereo(output, NUM_FRAMES, &a);
    print_analysis("Stereo Output", &a);

    free(input);
    free(output);
}

static void test_live_config_update() {
    printf("\n=== Test: Live Config Update ===\n");

    float* input = malloc(NUM_FRAMES * 6 * sizeof(float));
    float* output = malloc(NUM_FRAMES * 2 * sizeof(float));

    generate_tone(input, NUM_FRAMES, 2, 440.0f, 1.0f);

    spatial_downmix_config_t config;
    spatial_downmix_get_default_config_5_1(&config);

    spatial_downmix_ctx_t* ctx = spatial_downmix_create(&config);

    config.gains_5_1.center = 0.7f;
    spatial_downmix_update_config(ctx, &config);
    spatial_downmix_process(ctx, input, output, NUM_FRAMES);

    analysis_t a1;
    analyze_stereo(output, NUM_FRAMES, &a1);
    print_analysis("center=0.7", &a1);

    config.gains_5_1.center = 1.0f;
    spatial_downmix_update_config(ctx, &config);
    spatial_downmix_process(ctx, input, output, NUM_FRAMES);

    analysis_t a2;
    analyze_stereo(output, NUM_FRAMES, &a2);
    print_analysis("center=1.0", &a2);

    config.gains_5_1.center = 1.3f;
    spatial_downmix_update_config(ctx, &config);
    spatial_downmix_process(ctx, input, output, NUM_FRAMES);

    analysis_t a3;
    analyze_stereo(output, NUM_FRAMES, &a3);
    print_analysis("center=1.3", &a3);

    float ratio_1_0 = a2.rms_left / a1.rms_left;
    float expected = 1.0f / 0.7f;
    printf("Gain ratio (1.0/0.7): measured=%.4f expected=%.4f diff=%.4f\n",
           ratio_1_0, expected, fabsf(ratio_1_0 - expected));

    spatial_downmix_destroy(ctx);
    free(input);
    free(output);
}

int main() {
    printf("========================================\n");
    printf("Spatial Downmix Engine - Standalone Test\n");
    printf("========================================\n");

    test_5_1_defaults();
    test_stereo_passthrough();

    test_isolated_channel_5_1(0, "FL (Front Left)", 1.0f);
    test_isolated_channel_5_1(1, "FR (Front Right)", 1.0f);
    test_isolated_channel_5_1(2, "C  (Center)", 1.0f);
    test_isolated_channel_5_1(3, "LFE", 1.0f);
    test_isolated_channel_5_1(4, "SL (Side Left)", 1.0f);
    test_isolated_channel_5_1(5, "SR (Side Right)", 1.0f);

    test_live_config_update();

    printf("\n=== All Tests Complete ===\n");
    return 0;
}